src/canaries/src/Loops.cpp

```diff
@@ -0,0 +1,160 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <canaries/Loops.hpp>
+
+#include <chrono>
+#include <optional>
+
+#include <gennylib/Orchestrator.hpp>
+#include <gennylib/PhaseLoop.hpp>
+
+#include <metrics/MetricsReporter.hpp>
+#include <metrics/metrics.hpp>
+
+#if defined(__APPLE__)
+
+#include <mach/mach_time.h>
+
+#endif
+
+/**
+ * Use `rdtsc` as a low overhead, high resolution clock.
+ *
+ * Adapted from Google Benchmark
+ * https://github.com/google/benchmark/blob/439d6b1c2a6da5cb6adc4c4dfc555af235722396/src/cycleclock.h#L61
+ */
+inline int64_t now() {
+#if defined(__APPLE__)
+    return mach_absolute_time();
+#elif defined(__x86_64__) || defined(__amd64__)
+    uint64_t low, high;
+    __asm__ volatile("rdtsc" : "=a"(low), "=d"(high));
+    return (high << 32) | low;
+#endif
+}
+
+/**
+ * Do something in each iteration of the loop.
+ *
+ * @tparam WithPing Template to dictate what to do. Right now there are only
+ *                  two options, but this will be expanded over time to
+ *                  support more action types.
+ */
+template <bool WithPing>
+inline void doPingIfNeeded() {
+    if constexpr (WithPing) {
+        std::cout << "PING!" << std::endl;
+    } else {
+        int x = 0;
+        asm volatile("" : : "r,m"(x++) : "memory");
+    }
+}
+
+genny::TimeSpec operator""_ts(unsigned long long v) {
+    return genny::TimeSpec(std::chrono::milliseconds{v});
+}
+
+namespace genny::canaries {
+
+template <bool WithPing>
+int64_t Loops<WithPing>::simpleLoop() {
+    int64_t before = now();
+    for (int i = 0; i < _iterations; i++) {
+        doPingIfNeeded<WithPing>();
+    }
+    int64_t after = now();
+
+    return after - before;
+}
+
+template <bool WithPing>
+int64_t Loops<WithPing>::phaseLoop() {
+
+    Orchestrator o{};
+    v1::ActorPhase<int> loop{
+        o,
+        std::make_unique<v1::IterationChecker>(std::nullopt,
+                                               std::make_optional(IntegerSpec(_iterations)),
+                                               false,
+                                               0_ts,
+                                               0_ts,
+                                               std::nullopt),
+        1};
+
+    int64_t before = now();
+    for (auto _ : loop)
+        doPingIfNeeded<WithPing>();
+    int64_t after = now();
+
+    return after - before;
+}
+
+template <bool WithPing>
+int64_t Loops<WithPing>::metricsLoop() {
+
+    auto metrics = genny::metrics::Registry{};
+    metrics::Reporter reporter(metrics);
+
+    auto dummyOp = metrics.operation("metricsLoop", WithPing ? "db.ping()" : "Nop", 0u);
```

Comment:
There's a requirement in the ticket to report the result of this metric:
> Report the added mean and variance latency added within #2, #3, and #4

I'm wondering if this is still something we need.

We can get a sense of the mean by comparing the `phase-loop` and `real-loop` benchmarks. `variance` is likely very high for no-op loops because of CPU scheduling. It's likely more consistent but noisy for heavy loops like `db.ping()`, because there's no way to distinguish if the variance is coming from `db.ping()` or metrics internals.

Reply:
Seems fine to drop this requirement for now